### PR TITLE
Add banned API analyzer

### LINKED ISSUE #3777: Add analyzers to catch incorrect usage within tree
I see the same bugs over and over again. We should write tests or analyzers to catch them so that new devs don't run into them:

- Use IProjectAccessor over IProjectLockService. This avoids skipping:
  - Checkout from source control on write 
  - Call to ReleaseAsync to avoid blocking a thread pool thread on release of the lock

- Use/centralized usage of MSBuild construction model. This makes sure that we:
  - Escape/Unescape values
  - Implement the same behavior of walking, looking for properties, items, etc.

- Use ISafeProjectGuidService vs IProjectGuidService, IProjectGuidService2. This avoids:
  - Reading the GUID before we've been set it in the cases where the solution gives us a new one

- Use IUnconfiguredProjectTasksService.ProjectLoadInHost vs IProjectAsyncLoadDashboard.ProjectLoadInHost. This handles:
  - The situation where the project is closed before it has finished loading, instead of waiting forever.

- Use IUnconfiguredProjectTasksService.LoadedProjectAsync vs CommonProjectSystemTools.LoadedProject<T>.
  - This interface is unit testable.

- Use IProjectDynamicLoadComponent vs [ProjectAutoLoad] and [ConfiguredProjectAutoLoad]. This avoids accidently reloading the project while loading, switching branches, installing NuGet package, etc.

- Reading TargetFramework/TargetFrameworks properties. Only the configuration service should read these properties. Everyone else should be looking at "isVariantDimension" dimensions, or TargetFrameworkMoniker - and not know anything about these properties.

### PR DESCRIPTION
**Pull request by drewnoakes**
Fixes #3777.

Adds a new Roslyn analyzer that allows us to emit errors on usages of types/members/etc we do not want to use in the solution.

This analyzer was written by @Pilchie in dotnet/roslyn-analyzers#1781 and I finished up a few bits with help from @mavasani.

@jmarolf helped with changes to the package references, which fixes our part of dotnet/roslyn-analyzers#1750, and showing me where the CI builds end up.

Earlier commits fix/suppress errors that would otherwise surface due to later commits, so hopefully every commit in the PR builds correctly.

047b1f3f3b1bbf6b4f3862e4a443745fe408a9f2 suppresses errors for current usages of banned APIs using `#pragma warning`.

6b3cc19232ff1d8f402763aea5cc6a9c8f04904d removes a bunch of `PublicAPI.Shipped.txt` lines that the latest public API analyzer complained about. I think that change is correct, but it'd be good if @tmeschter and others could take a look.

There are also three diagnostics in `Microsoft.NetFramework.Analyzers` that I've suppressed but that we may wish to address in future.

I'll create new issues to track removing existing suppressed usages of banned symbols, and enabling `Microsoft.NetFramework.Analyzers` diagnostics.

### CODE REVIEW

File: build/Default.ruleset [hunk 1]
```diff
@@ -46,6 +46,7 @@
     <Rule Id="CA1721" Action="None" />
     <Rule Id="CA1724" Action="None" />
     <Rule Id="CA1815" Action="None" />
+    <Rule Id="CA1820" Action="None" />
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply build/Default.ruleset]: #4377

File: build/Default.ruleset [hunk 2]
```diff
@@ -57,6 +58,16 @@
...
   </Rules>
+  <Rules AnalyzerId="Microsoft.CodeQuality.Analyzers.Exp" RuleNamespace="Microsoft.CodeQuality.Analyzers.Exp">
+    <Rule Id="CA1062" Action="None" />
+    <Rule Id="CA1303" Action="None" />
+    <Rule Id="CA1508" Action="None" /> <!-- disable until 2.9.x per https://github.com/dotnet/roslyn-analyzers/issues/1921#issuecomment-446702992 -->
+  </Rules>
+  <Rules AnalyzerId="Microsoft.NetFramework.Analyzers" RuleNamespace="Microsoft.NetFramework.Analyzers">
+    <Rule Id="CA2153" Action="None" /> <!-- TODO do not catch CorruptedStateExceptions -->
+    <Rule Id="CA2235" Action="None" /> <!-- TODO mark all non-serializable fields -->
+    <Rule Id="CA3075" Action="None" /> <!-- TODO insecure DTD processing in XML -->
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply build/Default.ruleset]: I included these in #4377. Would you want them to be addressed in this PR?

File: build/Default.ruleset (hunk 1 above)
[the remark this answers is not in the document]
Comment (jmarolf) [reply build/Default.ruleset]: 👍

**Comment by jmarolf** [comment]
Also want to point out that this PR will move us a version of the compiler that has the nullable references feature enabled.  We could theoretically turn this on for the team

**Comment by Pilchie** [comment]
@jmarolf not really, due to the WPF project stuff I've been wrestling with all day.

File: build/Default.ruleset (hunk 2 above)
[the remark this answers is not in the document]
Comment (drewnoakes) [reply build/Default.ruleset]: I feel like they should be reviewed in a separate PR. They're not simple mechanical fixes, and most of the violations are in VB code.

[the remark this answers is not in the document]
Comment (Pilchie) [reply build/Default.ruleset]: I'd prefer if we fix the `CA3075` violations.

[the remark this answers is not in the document]
Comment (drewnoakes) [reply build/Default.ruleset]: Agreed via email to file another PR immediately.

**Pull request merged**